internal/pkg/composable/providers/kubernetesleaderelection/kubernetes_leaderelection_test.go

```diff
@@ -0,0 +1,191 @@
+// 4. We will force the lease to be acquired by the agent1 again. To avoid the agent2 reacquiring it multiple times,
+// we will stop this provider and make sure the agent1 can reacquire it.
+func TestNewLeaderElectionManager(t *testing.T) {
+	client := k8sfake.NewSimpleClientset()
+
+	lease := createLease()
+	// create the lease that leader election will be using
+	err := applyLease(client, lease, true)
+	require.NoError(t, err)
+
+	// Create the provider
+	logger := logp.NewLogger("test_leaderelection")
+
+	leaseDuration := 3
+	leaseRenewDeadline := 2
+	leaseRetryPeriod := 1
+
+	c := map[string]interface{}{
+		"leader_lease":         leaseName,
+		"leader_leaseduration": leaseDuration,
+		"leader_renewdeadline": leaseRenewDeadline,
+		"leader_retryperiod":   leaseRetryPeriod,
+	}
+	cfg, err := config.NewConfigFrom(c)
+	require.NoError(t, err)
+
+	getK8sClientFunc = func(kubeconfig string, opt autodiscoverK8s.KubeClientOptions) (kubernetes.Interface, error) {
+		return client, nil
+	}
+	require.NoError(t, err)
+
+	podNames := [2]string{"agent1", "agent2"}
+	cancelFuncs := [2]context.CancelFunc{}
+
+	// Create two leader election providers representing two agents running
+	for i := 0; i < 2; i++ {
+		p, err := ContextProviderBuilder(logger, cfg, true)
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		cancelFuncs[i] = cancel
+		comm := ctesting.NewContextComm(ctx)
+
+		err = os.Setenv("POD_NAME", podNames[i])
+		if err != nil {
+			require.FailNow(t, "Failed to set pod name environment variable.")
+		}
+		go func() {
+			_ = p.Run(ctx, comm)
+		}()
+
+		if i == 1 {
+			break
+		}
+
+		// We need to wait for the first agent to acquire the lease, so we can POD_NAME environment variable again
+		expectedLeader := leaderElectorPrefix + podNames[i]
+		for {
+			holder, err := getLeaseHolder(client)
+			require.NoError(t, err)
+
+			if holder == expectedLeader {
+				break
+			}
+		}
+	}
+
+	// At this point the current holder is agent1. Let's change it to agent2.
+	for {
+		// Force the lease to be applied again, so a new leader is elected.
+		intermediateHolder := "does-not-matter"
+		lease.Spec.HolderIdentity = &intermediateHolder
+		err = applyLease(client, lease, false)
+		require.NoError(t, err)
+
+		var currentHolder string
+		for {
+			currentHolder, err = getLeaseHolder(client)
+			require.NoError(t, err)
+
+			// In this case, we already have an agent as holder
+			if currentHolder == leaderElectorPrefix+podNames[0] || currentHolder == leaderElectorPrefix+podNames[1] {
+				break
+			}
+		}
+
+		if currentHolder == leaderElectorPrefix+podNames[1] {
+			break
+		}
+	}
+
+	// Now that the holder is agent2, let's wait for agent1 to be reelected.
+	// To avoid having to wait very long, the context of agent2 will be canceled so the leader elector will not be
+	// running anymore. This way there is only one instance fighting to acquire the lease.
+	cancelFuncs[1]()
```

Comment:
You can still have this line here if you add a `defer cancel()` above, the defer will just become a noop.